arrow/memory/mallocator/mallocator.go

```diff
@@ -58,28 +70,44 @@ func (alloc *Mallocator) Allocate(size int) []byte {
 	if size < 0 {
 		panic("mallocator: negative size")
 	}
-	ptr, err := C.calloc(C.size_t(size), 1)
+	paddedSize := C.size_t(size) + C.size_t(alloc.alignment)
+	ptr, err := C.calloc(paddedSize, 1)
 	if err != nil {
 		// under some circumstances and allocation patterns, we can end up in a scenario
 		// where for some reason calloc return ENOMEM even though there is definitely memory
 		// available for use. So we attempt to fallback to simply doing malloc + memset in
 		// this case. If malloc returns a nil pointer, then we know we're out of memory
 		// and will surface the error.
-		if ptr = C.malloc(C.size_t(size)); ptr == nil {
+		if ptr = C.malloc(paddedSize); ptr == nil {
 			panic(err)
 		}
-		C.memset(ptr, 0, C.size_t(size))
+		C.memset(ptr, 0, C.size_t(paddedSize))
 	} else if ptr == nil {
 		panic("mallocator: out of memory")
 	}
 
+	buf := unsafe.Slice((*byte)(ptr), paddedSize)
+	aligned := roundToPowerOf2(uintptr(ptr), uintptr(alloc.alignment))
+	alloc.realAllocations.Store(aligned, uintptr(ptr))
 	atomic.AddUint64(&alloc.allocatedBytes, uint64(size))
-	return unsafe.Slice((*byte)(ptr), size)
+
+	if uintptr(ptr) != aligned {
+		shift := aligned - uintptr(ptr)
+		return buf[shift : uintptr(size)+shift : uintptr(size)+shift]
```

Comment:
should the capacity still just be `size`?

Reply:
The third slice parameter is an index and not a capacity, so this sets the capacity to `size`.

Reply:
> The primary expression
> 
> a[low : high : max]
> 
> constructs a slice of the same type, and with the same length and elements as the simple slice expression a[low : high]. Additionally, it controls the resulting slice's capacity by setting it to max - low.

Reply:
Ah weird. I always thought it was an absolute capacity for some reason...